### DISCUSSION TITLE
2024-03-08 - Modern Stateless Authentication - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by Dhruv-0987** [SSWConsulting/SSW.Rules.Content#8074] [Chewing the Fat]
Hey SSW'ers,

Let's talk about stateless authentication.

https://www.ssw.com.au/rules/modern-stateless-authentication/

**Comment by leahy268** [comment]
Interesting topic.  I understood a fair bit if how that all worked before today. Nice way to clear the whole story up though.

**Comment by matt-goldman** [comment]
Great topic, looking forward to diving into it more. The CTF form took a bit of an incoherent turn toward IdentityServer. I understand the link but this wasn't clear.

For anyone wondering - Duende IdentityServer (formerly just IdentityServer but it's now a commercial product) is an OIDC implementation in .NET. It builds upon ASP.NET Core technology, which already handles things like user stores and session based auth, to provide JWT auth. It handles issuing and validating JWTs and everything that goes with it.

At SSW, we have implemented Duende IdentityServer as our central Single Sign-On (SSO) identity provider (IDP). SSO means using the same identity to sign in to multiple applications. How this ties to this week's CTF is that we can now use a single identity (think the same UN+PW combo) to sign in to all SSW applications (this is WIP as you've seen from various SSW all emails), using modern stateless auth with JWTs.

Sorry for the big mansplain, just thought it wasn't clear in the form, especially for non-devs or those who haven't worked with auth.

**Reply by brydeno** [reply]
And what a nice improvement to both security and user experience changing to SSO has been.

**Comment by 2fernandez** [comment]
nice video.

**Reply by matt-goldman** [reply]
Seen a couple of videos from this guy recently. They are very good.

**Comment by drwharris** [comment]
Great video. I am dealing with the issue of refresh tokens and the problems around their "Long livedness", lack of standard related to the use of refresh tokens and protecting form MITM attacks. IdentityServer and Auth0 give you mechanisms for dealing with it but B2C does not.

**Comment by Levijj22** [comment]
Interesting topic, he does a great job explaining it!

**Comment by joshbermanssw** [comment]
Havent delved too much into security before, i found the video quite informative (have been seeing alot of ByteByteGo recently). I found that the video and questions also sparked personal investigation into a few of the topics mentioned this week in CTF, a style i like!

**Comment by tkapa** [comment]
I've previously implemented IdentityServer as a junior dev (Mistake) but in the process, learned a ton about this process and how it works. Good to know I still understand it, for the most part.

**Comment by bettybondoc** [comment]
As a visual learner, I find ByteByteGo's videos especially appealing. Their clear, succinct explanations combined with engaging animated graphics make every video highly effective. My uni professors tried (and failed) so hard to explain this week's and last week's topics to me but this guy did it in 5 minutes. LOL

**Comment by cvionnet** [comment]
An important topic that can be tough to swallow, but the video and rules help to understand it better.
Having the pros and cons on each solution facilitates the process of selecting the best solution.
The video and rule are excellent!

**Comment by AlexRothwell** [comment]
This was a great topic to cover, it's an important concept to understand for any developer working on web applications

**Comment by duanxinhuan** [comment]
good to revisit what I have learned on 2FA

**Comment by TobyChurches** [comment]
This was a great CTF, really enjoyed the content and video! Authentication is yet another topic that wasn't taught well at Uni! Id love to see more of these Auth videos

**Comment by jeoffreyfischer** [comment]
Great topic, I never heard of JWT before - the whole process is still a bit abstract but I'll dig more into it!

**Comment by AntPolkanov** [comment]
I like when CTFs discuss a well-known topic and make it clearer or a topic which you knew long time ago and already forgot.

**Comment by awaltos** [comment]
JWT is a great idea provided that all the security considerations are taken into account and risks mitigated. I like how you can add your own custom claims in your auth provider which can be used as information client-side.
I had some experience using this system, but the video does a great job of explaining details and the overall flow. There is still plenty to deep dive into on this topic.

**Comment by chrisschultzssw** [comment]
Both the video and the rule explained this really well. 

In case anyone else was wondering...

> Pronunciation Quirk:
> The pronunciation “jot” is a play on words, as it sounds like the English word “jot”, which means a small or insignificant amount.
> By using this pronunciation, developers add a touch of whimsy to a technical term, making it more memorable and approachable.
> 
> Fun Fact:
> Some people find the formal pronunciation amusing, while others prefer to stick to the three letters’ Dutch pronunciation: “yaywaytay”.
> 
> In summary, whether you say “jot” or “yaywaytay”, JWTs remain a powerful tool for secure data transmission in web applications! 
🤖

**Reply by danielmackay** [reply]
or "JEWT" ;-)

**Comment by danielmackay** [comment]
The most important thing I learned today is that JWT is pronounced "JOT", not "JEWT". ;-)

Although, I've used JWT's in the past I hadn't thought about the pros and cons of using different algorithms for the signature before.  Something to keep in mind.

**Reply by ricksu978** [reply]
I pronounce it "JWT" 😆

**Reply by matt-goldman** [reply]
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/19944129/ca08f635-61b0-4fb5-9287-3b002cee7ce0)

**Comment by ricksu978** [comment]
Identity Server offers the option to issue the tokens in two formats
1. JWT
2. Reference Token (aka Opaque Token)

Now we have rule on JWT, maybe we could also talk about Reference Token, how to validate it, comparison to JWT...etc.

**Comment by JackDevAU** [comment]
Loving these videos - maybe we should do something similar with some of our devs! 
Let's make a beginner series on topics like this one! (Focusing on MS)

**Comment by Marxoz** [comment]
Really enjoying these CTF's where you learn something new. We tend to use these technologies and features without always knowing why. CTF's like these fill in some gaps, even if you are familiar with the topic

**Comment by stephanfakossw** [comment]
Nice rule and video. I've learned something.

**Comment by bahjatalaadelssw** [comment]
Great video, and easy to follow. The animations make the information more accessible.

**Comment by AshAnil** [comment]
Nice video ⭐ 
It explained well

**Comment by jimmy1992abc** [comment]
Interesting topic today, I like these type of videos becuase I can learn a little bit without being a developer.

I do have questions on how web applications manage their credentials. As increasing exposure of massive sensitive information leakage recent years, it brings us the attention on how to manage these data properly. Instead of storing these data on the server, validate each login and having the risk of "take-over", this might be a good direction in the future.

**Comment by Harry-Ross** [comment]
Great video, and the rule is a super convenient reference for the benefits and drawbacks of JWTs. I've already spent a lot of time using JWTs so I was already familiar with most of the content, but was a good refresher.

**Comment by YalalCakirSSW** [comment]
Great rule. I've learned a lot. It's crucial to know best practices when it comes to security.

**Comment by Ben0189** [comment]
Great video 🔥

**Comment by AttackOnMorty** [comment]
Great video. We should try to do a series on a certain topic for CTF.

**Comment by CoffeeBugHub** [comment]
Great rule, love the topic.

**Comment by KahaMason** [comment]
Great video and rule 👍

**Comment by MattParkerDev** [comment]
JWTs are great - they make auth easy.

**Comment by SamBWagner** [comment]
Nice video. JWTs are radical.

**Comment by jakebayliss** [comment]
Great video again. Learned a few new pros and cons of JWTs :)

**Comment by amankumarrr** [comment]
Great insight into JWTs for authentication and authorization

**Comment by ozair-ashfaque** [comment]
Great video and rule.

**Comment by jernejk** [comment]
I find the video great refresher and something I would share with others if they don't understand JWT. :)

I do see an issue regarding security, how we store it, invalidate it, what should be stored, etc. I can still say that it's a very heated conversation with no clear winner as of now; most seem to prefer local/session storage over cookies. I think the security varies between applications; a pure API can't execute injected JS, for instance, and cookies can't be easily compromised by something sniffing the browser's local/session storage.

My bias is with Blazor as it gives you secure storage for JWT and secure cookies managed by 3rd identity server.

|  	| Local/Session Storage | 	Cookies (Storage) |
| --- | --- | --- |
| JavaScript | Accessible through JavaScript on the same domain | Cookies, when used with the HttpOnly cookie flag, are not accessible through JavaScript |
| XSS attacks | Vulnerable to XSS attacks | Immune to XSS (with HttpOnly flag) |
| CSRF attacks | Immune to CSRF attacks | Vulnerable to CSRF attacks |
|Mitigation | Do not store private/sensitive/authentication-related data here | Make use of CSRF tokens or other prevention methods |

Source: https://jerrynsh.com/all-to-know-about-auth-and-cookies/


In the rule, we are recommending "Prefer HttpOnly cookies over local storage for JWTs to mitigate the risk of Cross-Site Scripting (XSS) attacks; HttpOnly cookies enhance security by restricting access to tokens via JavaScript".

If we go with this recommendation to mitigate XSS, we should also expand on how to deal with CSRF as that is a topic that is not well understood.

We should also mention how to create an invalidation process. There are 2 major ways to do it, to my knowledge:

| Method | Description | ✅ Pro | ❌ Cons |
| --- | --- | --- | --- |
| Allow-list | Have a list of all allowed JWTs | Full control over access, great for long-lived tokens | Maintenance tasks, scalability and not great with short-lived tokens |
| Auth code/id | Have a code for each user which gets resets when signing out users similar to reset password | Efficient, "sign-out everywhere" feature, easy to implement | Can't revoke individual tokens |

**Reply by ricksu978** [reply]
CSRF can be mitigated by Same-Site cookies.

I would prefer Same-Site cookie over Anti-CSRF Token.
Just because IMO, it's easier to implement.

**Comment by jerryluossw** [comment]
Great rule, and I think the best use case for JWT is not Internet applications, but interface calls in microservice architectures or distributed architectures.

**Comment by Aibono1225** [comment]
Great rule and video, I had learned about JWT before, but now it's much clearer to me

**Comment by babakamyljanovssw** [comment]
Interesting topic, the video and rule explained very clear about JWT!

**Comment by SofieHong** [comment]
Great video, great explanation

**Comment by tombui99** [comment]
Great video, it's good for the devs to refresh their knowledge once in a while.

**Comment by KristenHu** [comment]
Great video, explain JWT clearly, I like this author👍

**Comment by jimmidier** [comment]
Great video and a well written rule!

**Comment by synnoveisabel** [comment]
Interesting and important topic, the video and rule gave a good explanation

**Comment by tomek-i** [comment]
Good topic and video. It would have been great to combine it with session based auth to illustrate the differences between them

**Comment by steven0x51** [comment]
Great vido, have a better insight of JWT

**Comment by jackreimers** [comment]
Very informative rule and video. It's very important to understand these concepts so we can build safe and secure applications.

**Comment by tino-liu** [comment]
Very informative. I know that JWT is encoded but never thought about how to use it safely and securely. This video gave a good explanation.

**Comment by PothieuG** [comment]
Thanks for the reminder !
Always good see how JWT works with greats images/graphics !

**Comment by Freego1783** [comment]
Important topic!

I haven't really worked on the security aspect of applications, but I am familiar with the basic principles. I should delve into it more!

**Comment by camillars** [comment]
Interesting video - the visuals make it easier to understand the concept.

**Comment by Calinator444** [comment]
Interesting video.

Although if the header of a JWT describes the encryption algorithm used but the payload doesn't need to be encrypted what does the header describe when the payload isn't encrypted?

**Reply by matt-goldman** [reply]
This is a really good question! It's actually the signing algorithm, not the encryption algorithm (although technically the same, it's used in a JWT to sign rather than encrypt)/

**Comment by christoment** [comment]
Good topic for people to learn stateless authentication 😀

**Comment by vladislav-kir** [comment]
Great topic that refreshed my memory on JWT tokens.

**Comment by PennyWalker** [comment]
I really like the style of this guys videos, they are so easy to understand and the animations are really helpful. I learned a lot.
